File: 35.复杂链表的复制/35.复杂链表的复制.py
# -*- coding:utf-8 -*-
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None
class Solution:
    # 返回 RandomListNode
    def Clone(self, pHead):
        # write code here
        if pHead is None:
            return None
        hashset = dict()
        p = pHead
        helpHead = RandomListNode(-1)
        q = helpHead
        while(p is not None):
            newNode = RandomListNode(p.label)
            hashset[p] = newNode 
            q.next = newNode
            q = q.next
            p = p.next
        p = pHead
        q = helpHead.next
        while(p is not None):
            q.random = hashset.get(p.random,None)
            p = p.next
            q = q.next
        return helpHead.next
            

File: 35.复杂链表的复制/35.复杂链表的复制.cpp
/*
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(NULL), random(NULL) {
    }
};
*/
class Solution {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        if(pHead == nullptr)
            return nullptr;
        RandomListNode* p = pHead,*temp =nullptr;
        while(p != nullptr){
            RandomListNode *q = new RandomListNode(p->label);
            temp = p->next;
            p->next = q;
            q->next = temp;
            p = temp;
        }
        p = pHead;
        while(p!=nullptr){
            if (p->random != nullptr){
                p->next->random = p->random->next;
            }
            p = p->next->next;
        }
        RandomListNode *newHead = new RandomListNode(-1);
        RandomListNode *tail = newHead;
        p = pHead;
        while(p != nullptr){
            temp = p->next->next;
            tail->next = p->next;
            p->next = temp;
            p = p->next;
            tail = tail->next;
        }
        return newHead->next;
    }
};